kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java

```diff
@@ -47,14 +47,16 @@ public static Table forPath(TableClient tableClient, String path)
 
     private final Path logPath;
     private final Path dataPath;
+    private final SnapshotManager snapshotManager;
 
     public TableImpl(Path logPath, Path dataPath) {
         this.logPath = logPath;
         this.dataPath = dataPath;
+        this.snapshotManager = new SnapshotManager();
```

Comment:
One this with this change is, we have a state in `Table`. We need to worry about thread safety. For example one thread is loading the `Table.getLatestSnapshot` and other thread calling `Table.getSnapshotAtVersion`. 

We may need to mark the APIs as thread-safe or make sure there are no concurrency issues.

Reply:
What is the thread unsafe concern here? What's unsafe about calling `.getLatestSnapshot` and then calling `.getSnapshotAtVersion`?

When these snapshots register themselves as the latest snapshot hint with the SnapshotManager, we check their version.

BTW, would be great if we wanted to keep a fixed size-bounded cache of snapshot hints , eh?

Reply:
> What is the thread unsafe concern here?
See my other comment about this.

> would be great if we wanted to keep a fixed size-bounded cache of snapshot hints , eh?
At some point we have definitely wandered into the anti-goal of complexity in the name of optimizing performance. 

The actual goal  is to be as simple as possible while not pessimizing performance. If we really want to allow optimization here, we should probably allow the caller to pass in the hint directly (from whatever snapshot they like), rather than taking an opinion about how many snapshots/hints to remember internally.

Reply:
Good catch... we've been intentionally avoiding internal state like this so far.

And it's especially messy because the snapshot manager is captured by the snapshot as well, but only inside a lazy value that will make it ~impossible to guess lifetimes.

Following the other comment -- if we accept that creating a snapshot _will_ access the table at least enough to compute P&M, then we can get rid of at least some of this messiness by having `SnapshotManager.createSnapshot` set the hint directly:
```java
private SnapshotImpl createSnapshot(...) {
    ...
  final val snapshot = new SnapshotImpl(...);
  final SnapshotHint hint = new SnapshotHint(
    snapshot.getVersion(tableClient), snapshot.getProtocol(), snapshot.getMetadata());
  registerHint(hint);
  return snapshot;
}
```
(something like the above would get rid of two lazy values and remove the need to pass the snapshot manager into the snapshot constructor)

Reply:
What is you want to construct a LogReplay but not get the protocol and metadata?

Reply:
I'd prefer to see a compelling example of that use case before optimizing for it. But LogReplay is anyway internal, so it's really "what if you want to construct a Snapshot but not get P&M?"

We don't want to eagerly grab app ids because there's no guarantee the user will access it.
We don't eagerly replay file actions because pruning will hopefully eliminate most of them.

But it's impossible to interact meaningfully with a Delta table without knowing P&M -- all you can learn without them is the latest commit version, and we can get that from the log segment without constructing a LogReplay at all.

Reply:
I will make all this not `lazy` in a followup pr

Reply:
Can we please get rid of the lazy now, not in a future PR? It's weird to be changing the API and adding complexity, just to accommodate something should be removed.